nilearn/decomposition/base.py

```diff
@@ -0,0 +1,451 @@
+"""
+Base class for decomposition estimators, utilities for masking and dimension
+reduction of group data
+"""
+from __future__ import division
+
+from math import ceil
+
+import numpy as np
+from scipy import linalg
+from sklearn.base import BaseEstimator
+from sklearn.externals.joblib import Memory, Parallel, delayed
+from sklearn.linear_model import LinearRegression
+from sklearn.utils import check_random_state
+from sklearn.utils.extmath import randomized_svd
+
+from .._utils.cache_mixin import CacheMixin, cache
+from .._utils.niimg import _safe_get_data
+from .._utils.niimg_conversions import check_niimg_4d
+from ..input_data import NiftiMapsMasker
+from ..input_data.masker_validation import check_embedded_nifti_masker
+
+
+def mask_and_reduce(masker, imgs,
+                    confounds=None,
+                    reduction_ratio='auto',
+                    n_components=None, random_state=None,
+                    memory_level=0,
+                    memory=Memory(cachedir=None),
+                    n_jobs=1):
+    """Mask and reduce provided 4D images with given masker.
+
+    Uses a PCA (randomized for small reduction ratio) or a range finding matrix
+    on time series to reduce data size in time direction. For multiple imageS,
+    the concatenation of data is returned, either as an ndarray or a memorymap
+    (useful for big datasets that do not fit in memory).
+
+    Parameters
+    ----------
+    masker: NiftiMasker or MultiNiftiMasker
+        Instance used to mask provided data.
+
+    imgs: list of Niimg-like objects
+        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+        List of subject data to mask, reduce and stack.
+
+    confounds: CSV file path or 2D matrix, optional
+        This parameter is passed to signal.clean. Please see the
+        corresponding documentation for details.
+
+    reduction_ratio: 'auto' or float between 0. and 1.
```

Comment:
sorry, but kinda far from being clear what this does, when not considering the code